Extend CampaignsUsageStatistics with spec count/diff stat

mrnugget opened the pull request:
This fixes https://github.com/sourcegraph/sourcegraph/issues/15278 by adding multiple new fields to `types.CampaignsUsageStatistics`:

- The sum of the diff stats for changesets:
  - `ActionChangesetsDiffStatAddedSum`
  - `ActionChangesetsDiffStatChangedSum`
  - `ActionChangesetsDiffStatDeletedSum`
- The sum of the diff stats for _merged_ changesets:
  - `ActionChangesetsMergedDiffStatAddedSum`
  - `ActionChangesetsMergedDiffStatChangedSum`
  - `ActionChangesetsMergedDiffStatDeletedSum`
- `CampaignSpecsCreatedCount` - the number of created campaign specs (based on `event_logs`, which means historical data won't be available)
- `ChangesetSpecsCreatedCount` - the number of created changeset specs that were then used to create a campaign spec (based on the same `event_logs`)
- `CampaignsClosedCount` - the number of closed campaigns

The diff stat fields are queried with more clauses in the existing `SELECT` query. For the spec counts I had to add event logging to the `campaigns/resolver.Resolver` to insert an `usagestats.Event` whenever a user created a campaign spec. I decided to keep that in the resolver instead of adding it to our `enterprise/campaigns` package, since I saw that other event logging in `frontend` is also done on the resolver level.

PR to change BigQuery schema: https://github.com/sourcegraph/analytics/pull/67

---

@ebrodymoore Two things:
1. Can you review this?
2. I understand that this requires adding these new fields to the BigQuery schema — how do we do that? Do I do that or do you?


---

Why do we collect this data?

To use it as a metric for our goals in Q4 2020. We want to increase the usage of campaigns and the amount of code that's changed with campaigns. In order to measure our performance, we need to collect these numbers.

Since they're all either counts or the sum of counts, the data is anonymous and only gives us insight into how much campaigns are used, not by who or for what.

chrispine commented:
This is really cool! Thanks 😄

eseliger commented on enterprise/internal/campaigns/service.go:
```diff
@@ -46,12 +46,12 @@ func (s *Service) WithStore(store *Store) *Service {
 
 type CreateCampaignSpecOpts struct {
-	RawSpec string
+	RawSpec string `json:"raw_spec"`
```
Do you think we should replicate the entire spec here again? This might bloat the table even further

eseliger replied:
Maybe just saving the one datapoint that we actually care about would be enough? (The length of the rand ids)

mrnugget replied:
Good point. But I'm not sure, honestly. I assumed we add it to the table so we can query number of steps, etc. in the future. And compared to the number of other events, I assume these ones to be just a drop in the bucket (average size of a campaign spec YAML is ~1kb in my specs folder).

eseliger replied:
I guess once we want to track that, we could still add more fields? I know we can't look in the past then, but I'm still a little hesitant to persist so much data for so few used fields of it. Maybe I'm just overcautious here, but it feels like we should not just add 1kb per row, "because others do it as well".

mrnugget replied:
Alright, alright, I'll change it :)

eseliger reviewed (approved):
Cool, LGTM. One question about efficiency and the size of the already pretty large event_logs table.

mrnugget commented:
Heads up: I updated the PR description to include "Why", a link to the analytics PR (https://github.com/sourcegraph/analytics/pull/67), added an entry to the changelog and updated both "pings" pages (in docs and site-admin)

LawnGnome commented on internal/usagestats/campaigns.go:
```diff
@@ -9,38 +9,51 @@ import (
     (SELECT COUNT(*) FROM campaigns) AS campaigns_count,
-    COUNT(*) FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED') AS action_changesets,
-    COUNT(*) FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED' AND external_state = 'MERGED') AS action_changesets_merged,
+    COUNT(*)                        FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED') AS action_changesets,
+    COALESCE(SUM(diff_stat_added)   FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED'), 0) AS action_changesets_diff_stat_added_sum,
+    COALESCE(SUM(diff_stat_changed) FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED'), 0) AS action_changesets_diff_stat_changed_sum,
+    COALESCE(SUM(diff_stat_deleted) FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED'), 0) AS action_changesets_diff_stat_deleted_sum,
+    COUNT(*)                        FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED' AND external_state = 'MERGED') AS action_changesets_merged,
+    COALESCE(SUM(diff_stat_added)   FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED' AND external_state = 'MERGED'), 0) AS action_changesets_merged_diff_stat_added_sum,
+    COALESCE(SUM(diff_stat_changed) FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED' AND external_state = 'MERGED'), 0) AS action_changesets_merged_diff_stat_changed_sum,
+    COALESCE(SUM(diff_stat_deleted) FILTER (WHERE owned_by_campaign_id IS NOT NULL AND publication_state = 'PUBLISHED' AND external_state = 'MERGED'), 0) AS action_changesets_merged_diff_stat_deleted_sum,
```
This is the kind of SQL I really enjoy, particularly when someone who's not me has written it.

mrnugget replied:
It's all about that aligned indentation!

ebrodymoore reviewed (approved):
Looks good to me from my end - I trust that you and Erik can handle the review on the code side.

Two other things:
- I added @dadlerj as a reviewer so he can okay that this data goes into pings. The purpose is good and the data to be collected is straightforward, but want to make sure he checks off on it.
- I'm testing sourcegraph/analytics#67 over the next 24 hours so please hold off on merging this until I give you the go-ahead (assuming Dan is cool with it!).

LawnGnome reviewed (approved):
I share Erik's concern around replicating the entire spec, but otherwise this looks like it has what we talked about and want. :rocket:

mrnugget commented:
Code is updated to now only save the number of changeset specs in the `event.argument` field.

Now waiting for @ebrodymoore and @dadlerj to approve.

ebrodymoore commented:
I've made the updates to the schema so as long as Dan checks off on this, it's ready to be merged

mrnugget commented:
@ebrodymoore @dadlerj I just pushed another commit, 237b81d, that adds another field: `CampaignsClosedCount`. That is required in #15482 and I don't want to have two PRs, two changelog entries, etc. just to add another field, so I added it here.

The corresponding analytics PR is here https://github.com/sourcegraph/analytics/pull/71.

@ebrodymoore can you review the newest commit and hit approve if everything's okay and also look at https://github.com/sourcegraph/analytics/pull/71?
@dadlerj can you review this here?

dadlerj commented on doc/admin/pings.md:
```diff
@@ -46,9 +46,14 @@ By default, Sourcegraph also aggregates usage and performance metrics for some p
   - Total count of created campaigns
+  - Total count of closed campaigns
   - Total count of changesets created by campaigns
+  - Aggregate counts of lines changed, added, deleted in changeset
```
Is this in each changeset, or in aggregate? Wording here is a bit unclear (e.g. perhaps change it to "in all changesets")

mrnugget replied:
Fixed! Sorry about that, that was a typo/lazy-copy.

dadlerj commented on doc/admin/pings.md:
```diff
@@ -46,9 +46,14 @@ By default, Sourcegraph also aggregates usage and performance metrics for some p
   - Total count of changesets created by campaigns
+  - Aggregate counts of lines changed, added, deleted in changeset
   - Total count of changesets created by campaigns that have been merged
+  - Aggregate counts of lines changed, added, deleted in merged changeset
```
Same

dadlerj commented on enterprise/internal/campaigns/resolvers/resolver.go:
```diff
@@ -360,6 +366,29 @@ func (r *Resolver) CreateCampaignSpec(ctx context.Context, args *graphqlbackend.
+		return err
+	}
+
+	return usagestats.LogEvent(ctx, usagestats.Event{
```
We provide a convenience function, `logBackendEvent`, in the same file as `logEvent` for any events logged on the backend. Perhaps use that instead?

mrnugget replied:
You just made my day! Thanks!

dadlerj commented on enterprise/internal/campaigns/resolvers/resolver.go:
```diff
@@ -360,6 +366,29 @@ func (r *Resolver) CreateCampaignSpec(ctx context.Context, args *graphqlbackend.
+		EventName: "CampaignSpecCreated",
+		UserID:    actor.UID,
+		Source:    "backend",
+		Argument:  json.RawMessage(jsonArg),
```
Thank you for slimming this down :). BizOps team strongly, strongly, strongly prefers specific data only to be added here as opposed to arbitrary arguments that can be passed in (not just for table bloat reasons—in our old events db we used to see binary files accidentally get passed into the "args" column—but also to prevent  PII, etc. from accidentally landing in the events db).

mrnugget replied:
Now that you mentioned PII I realise that this _would've_ leaked authorName and authorEmail specified in a campaign spec. So, thanks again @eseliger for being table-row-size-aware :)

dadlerj reviewed (approved):
A few small notes but seems fine. Thanks for documenting/testing this well.

LawnGnome reviewed (approved):
Still LGTM.

mrnugget commented:
Heads up everyone:

1. I'm adding two more metrics to this after merging #15617, which adds two more events, and talking to @ebrodymoore about it. That will then fix https://github.com/sourcegraph/sourcegraph/issues/15482 completely.
2. I'll also add those two metrics to https://github.com/sourcegraph/analytics/pull/71

mrnugget commented:
Alright, this contains all the counts requires by #15482, fully documented.

https://github.com/sourcegraph/analytics/pull/71 has also been updated.

@ebrodymoore Waiting for final approval now before merging this.

ebrodymoore commented:
@mrnugget I'm reviewing this soon, please wait to merge until I give you the go ahead. Merging sourcegraph/analytics#71 doesn't put anything in prod, I still need to make the updates to the BigQuery schema.

ebrodymoore commented:
I've updated the prod schema so this is ready to be merged!
